File: solutions/121-best-time-to-buy-and-sell-stock/solution.py
from typing import List


class Solution:
    def max_profit(self, prices: List[int]) -> int:
        min_price = prices[0]
        max_profit = 0
        for price in prices:
            if price < min_price:
                min_price = price
            elif price - min_price > max_profit:
                max_profit = price - min_price
        return max_profit


if __name__ == "__main__":
    sol = Solution()
    got = sol.max_profit([7, 1, 5, 3, 6, 4])
    status = "PASS" if got == 5 else "FAIL"
    print(f"{status} | profit={got}")

File: solutions/121-best-time-to-buy-and-sell-stock/solution.go
package main

import "fmt"

func maxProfit(prices []int) int {
	minPrice := prices[0]
	maxProfit := 0
	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if price-minPrice > maxProfit {
			maxProfit = price - minPrice
		}
	}
	return maxProfit
}

func main() {
	got := maxProfit([]int{7, 1, 5, 3, 6, 4})
	status := "PASS"
	if got != 5 {
		status = "FAIL"
	}
	fmt.Printf("%s | profit=%d\n", status, got)
}
